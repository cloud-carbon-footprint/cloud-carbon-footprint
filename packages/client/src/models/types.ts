/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */
import { DropdownOption } from '../dashboard/filters/DropdownFilter'

export interface serviceEstimate {
  cloudProvider: string
  accountName: string
  serviceName: string
  kilowattHours: number
  co2e: number
  cost: number
  region: string
  usesAverageCPUConstant?: boolean
}

export interface ServiceResult<T> {
  data: T[]
  loading: boolean
}

export interface EstimationResult {
  timestamp: Date
  serviceEstimates: serviceEstimate[]
}

export interface cloudEstPerDay {
  x: Date
  y: number
  usesAverageCPUConstant?: boolean
  kilowattHours?: number
  cost?: number
}

export enum ChartDataTypes {
  REGION = 'region',
  SERVICE = 'service',
  ACCOUNT = 'account',
}

export interface FilterResultResponse {
  accounts: DropdownOption[]
  services: DropdownOption[]
}

export enum UnknownTypes {
  UNKNOWN_REGION = 'Unknown Region',
  UNKNOWN_SERVICE = 'Unknown Service',
  UNKNOWN_ACCOUNT = 'Unknown Account',
}

export interface EmissionsRatios {
  region: string
  mtPerKwHour: number
}
