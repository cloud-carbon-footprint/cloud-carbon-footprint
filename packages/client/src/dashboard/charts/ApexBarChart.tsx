/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { FunctionComponent, useState, Fragment } from 'react'
import { useTheme } from '@material-ui/core/styles'
import Chart from 'react-apexcharts'

import { sumCO2ByServiceOrRegion } from '../transformData'
import { ApexChartProps } from './common/ChartTypes'
import { EmissionsRatios, chartBarCustomColors } from '../../models/types'
import { Page, Pagination } from './Pagination'
import ChartLegend from './ChartLegend'
import NoDataPage from '../NoDataPage'
import useRemoteEmissionService from '../client/EmissionFactorServiceHook'

const mapToRange = (
  value: number,
  in_min: number,
  in_max: number,
  out_min: number,
  out_max: number,
) => {
  return ((value - in_min) * (out_max - out_min)) / (in_max - in_min) + out_min
}

export interface Entry {
  x: string
  y: number
}

export const ApexBarChart: FunctionComponent<ApexChartProps> = ({
  data,
  dataType,
}) => {
  const [pageData, setPageData] = useState<Page<Entry>>({ data: [], page: 0 })
  const theme = useTheme()

  const {
    data: emissionsData,
    loading: _emissionsLoading,
  } = useRemoteEmissionService()

  const mainTheme = theme.palette.primary.main
  const darkTheme = theme.palette.primary.dark

  let customBarColors = [mainTheme]
  if (dataType === 'region' && !!emissionsData.length) {
    customBarColors = createCustomBarColors(pageData, emissionsData, mainTheme)
  }

  const barChartData = sumCO2ByServiceOrRegion(data, dataType)

  const dataEntries: { x: string; y: number }[] = Object.entries(barChartData)
    .filter((item) => item[1] > 0)
    .map((item) => ({
      x: item[0],
      y: item[1],
    }))
    .sort((higherC02, lowerCO2) => lowerCO2.y - higherC02.y)

  const smallestCO2E = dataEntries?.[dataEntries?.length - 1]?.y
  const largestCO2E = dataEntries?.[0]?.y
  const totalCO2EByDataType = dataEntries.reduce((acc, currentValue) => {
    return acc + currentValue.y
  }, 0)

  const pageSize = 10
  const minThreshold = 1
  const maxThreshold = 100
  const mappedDataEntries: Entry[] = dataEntries.map((entry) => {
    const yEntry = mapToRange(
      entry.y,
      smallestCO2E,
      largestCO2E,
      minThreshold,
      maxThreshold,
    )
    return { x: entry.x, y: isNaN(yEntry) ? maxThreshold : yEntry }
  })

  const toolbarOffset = {
    x: -120,
    y: dataType === 'region' ? -122 : -55,
  }

  const options = {
    series: [
      {
        name: 'Total CO2e',
        data: pageData.data,
      },
    ],
    colors: customBarColors,
    chart: {
      type: 'bar',
      toolbar: {
        offsetX: toolbarOffset.x,
        offsetY: toolbarOffset.y,
        tools: {
          download: `
            <div class="apexcharts-menu-icon" title="Menu">
                <svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true">
                <path d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"></path>
                </svg>
            </div>   
           `,
        },
      },
    },
    grid: {
      show: false,
      yaxis: {
        lines: {
          show: false,
        },
      },
      xaxis: {
        lines: {
          show: false,
        },
      },
      padding: {
        left: 32,
      },
    },
    plotOptions: {
      bar: {
        horizontal: true,
        barHeight: `${7 * pageData.data.length}%`,
        distributed: true,
      },
    },
    legend: {
      show: false,
    },
    dataLabels: {
      enabled: true,
      textAnchor: 'start',
      formatter: function (_: number, opts: { dataPointIndex: number }) {
        const currentCO2E =
          dataEntries[pageData.page * pageSize + opts.dataPointIndex]?.y

        const formattedPercentage = (currentCO2E / totalCO2EByDataType) * 100
        return formattedPercentage < 0.01
          ? '< 0.01 %'
          : `${formattedPercentage.toFixed(2)} %`
      },
      offsetX: 16,
      background: {
        enabled: true,
        foreColor: mainTheme,
        borderColor: darkTheme,
        padding: 6,
        borderRadius: 1,
        borderWidth: 1,
        opacity: 0.9,
      },
    },
    xaxis: {
      type: 'category',
      labels: {
        show: false,
      },
      axisBorder: {
        show: false,
      },
      max: maxThreshold,
    },
    yaxis: {
      labels: {
        style: {
          fontSize: '13px',
        },
      },
    },
    tooltip: {
      fillSeriesColor: false,
      x: {
        show: false,
      },
      y: {
        formatter: function (value: number, opts: { dataPointIndex: number }) {
          return `${dataEntries[
            pageData.page * pageSize + opts.dataPointIndex
          ].y.toFixed(3)} metric tons`
        },
      },
    },
    height: '500px',
  }

  const handlePage = (page: Page<Entry>) => {
    setPageData(page)
  }

  return (
    <div
      style={{
        minHeight: '500px',
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'center',
        alignContent: 'center',
      }}
    >
      {emissionsData?.length ? (
        <Fragment>
          {dataType === 'region' && (
            <ChartLegend
              startLabel="Low carbon intensity"
              endLabel="High carbon intensity"
              colorRange={chartBarCustomColors}
            />
          )}
          <Chart
            options={options}
            series={options.series}
            type="bar"
            height={options.height}
          />
        </Fragment>
      ) : (
        <NoDataPage isTop={false} />
      )}
      <div style={{ paddingTop: '10px' }}>
        <Pagination
          data={mappedDataEntries}
          pageSize={pageSize}
          handlePage={handlePage}
        />
      </div>
    </div>
  )
}

export const createCustomBarColors = (
  pageData: Page<Entry>,
  emissionsData: EmissionsRatios[],
  mainTheme: string,
): string[] => {
  const regionColorsMap: string[] = []
  pageData.data.forEach((region) => {
    const currentRegion = region.x
    let color = chartBarCustomColors[0]
    const regionEmissionData = emissionsData.find(
      (item) => item.region === currentRegion,
    ) as EmissionsRatios
    if (!regionEmissionData) {
      regionColorsMap.push(mainTheme)
    } else {
      const { mtPerKwHour } = regionEmissionData
      if (mtPerKwHour >= 0.00064) {
        color = chartBarCustomColors[4]
      } else if (mtPerKwHour >= 0.00048 && mtPerKwHour < 0.00064) {
        color = chartBarCustomColors[3]
      } else if (mtPerKwHour >= 0.00032 && mtPerKwHour < 0.00048) {
        color = chartBarCustomColors[2]
      } else if (mtPerKwHour >= 0.00016 && mtPerKwHour < 0.00032) {
        color = chartBarCustomColors[1]
      } else {
        color = chartBarCustomColors[0]
      }
      regionColorsMap.push(color)
    }
  })
  return regionColorsMap
}
