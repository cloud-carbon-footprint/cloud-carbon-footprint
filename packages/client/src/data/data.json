[
    {
        "serviceEstimates": [
            {
                "co2e": 2,
                "cost": 0,
                "serviceName": "ebs",
                "timestamp": "2020-08-28T00: 00: 00.000Z",
                "wattHours": 0.5108360883835144
            },
            {
                "co2e": 1,
                "cost": 0,
                "serviceName": "s3",
                "timestamp": "2020-08-28T00: 00: 00.000Z",
                "wattHours": 0.00026440008612119945
            },
            {
                "co2e": 0,
                "cost": 2.330506638787581,
                "serviceName": "ec2",
                "timestamp": "2020-08-28T00: 00: 00.000Z",
                "wattHours": 50.5206029585591
            }
        ],
        "timestamp": "2020-08-28T00: 00: 00.000Z"
    },
    {
        "serviceEstimates": [
            {
                "co2e": 4,
                "cost": 0,
                "serviceName": "ebs",
                "timestamp": "2020-07-28T00: 00: 00.000Z",
                "wattHours": 0.23268313620749437
            },
            {
                "co2e": 4,
                "cost": 0,
                "serviceName": "s3",
                "timestamp": "2020-07-28T00: 00: 00.000Z",
                "wattHours": 0.0001680875709602132
            },
            {
                "co2e": 4,
                "cost": 1.9776640256171394,
                "serviceName": "ec2",
                "timestamp": "2020-07-28T00: 00: 00.000Z",
                "wattHours": 50.10627201163577
            }
        ],
        "timestamp": "2020-07-28T00: 00: 00.000Z"
    }
]