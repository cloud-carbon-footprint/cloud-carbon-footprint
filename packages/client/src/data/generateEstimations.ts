/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import moment from 'moment'
import { EstimationResult, serviceEstimate } from '../models/types'

const getRandomInt = (max: number): number => {
  return Math.floor(Math.random() * Math.floor(max))
}

const getRandomNumberInRange = (minValue: number, maxValue: number): number => {
  return Math.max(Math.random() * maxValue, minValue + Math.random())
}

const fakeServiceMap: () => { [key: string]: serviceEstimate } = () => {
  return {
    ebs: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 'ebs',
      kilowattHours: Math.random(),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      usesAverageCPUConstant: false,
    },
    s3: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 's3',
      kilowattHours: Math.random() / 1000,
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-2',
      usesAverageCPUConstant: false,
    },
    ec2: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 'ec2',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-west-1',
      usesAverageCPUConstant: false,
    },
    rds: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 'rds',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-west-2',
      usesAverageCPUConstant: false,
    },
    lambda: {
      cloudProvider: 'AWS',
      accountName: `aws account ${getRandomInt(5)}`,
      serviceName: 'lambda',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      usesAverageCPUConstant: false,
    },
    elasticache: {
      cloudProvider: 'AWS',
      accountName: `gcp account ${getRandomInt(5)}`,
      serviceName: 'elasticache',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east-1',
      usesAverageCPUConstant: false,
    },
    computeEngine: {
      cloudProvider: 'GCP',
      accountName: `gcp account ${getRandomInt(5)}`,
      serviceName: 'computeEngine',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'us-east1',
      usesAverageCPUConstant: false,
    },
    virtualMachines: {
      cloudProvider: 'AZURE',
      accountName: `azure account ${getRandomInt(5)}`,
      serviceName: 'virtualMachines',
      kilowattHours: getRandomNumberInRange(50, 75),
      co2e: getRandomInt(6),
      cost: getRandomNumberInRange(1.5, 2),
      region: 'UK South',
      usesAverageCPUConstant: false,
    },
  }
}

const generateEstimations = (
  today: moment.Moment | Date,
  monthsBack: number,
  servicesToTest: string[] = [
    'ebs',
    's3',
    'ec2',
    'rds',
    'lambda',
    'elasticache',
    'computeEngine',
    'virtualMachines',
  ],
): EstimationResult[] => {
  const todayAsMoment: moment.Moment = moment(today)

  const estimationResults: EstimationResult[] = []

  for (let i = 0; i <= monthsBack; i++) {
    const timestamp: Date = todayAsMoment
      .clone()
      .utc()
      .subtract(i, 'M')
      .hours(0)
      .minutes(0)
      .seconds(0)
      .millisecond(0)
      .toDate()
    const estimationsForMonth: EstimationResult = {
      timestamp,
      serviceEstimates: servicesToTest.map((serviceToTest: string) => {
        return fakeServiceMap()[serviceToTest]
      }),
    }

    estimationResults.push(estimationsForMonth)
  }

  return estimationResults
}

export default generateEstimations
