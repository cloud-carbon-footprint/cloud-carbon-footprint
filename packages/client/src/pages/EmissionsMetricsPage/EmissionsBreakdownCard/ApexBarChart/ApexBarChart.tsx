/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React, { FunctionComponent, useState } from 'react'
import Chart from 'react-apexcharts'
import { useTheme } from '@material-ui/core/styles'
import { PageEntry, Page, barChartCustomColors } from '../../../../Types'
import useStyles from './apexBarChartStyles'
import { createCustomBarColors, mapToRange } from './helpers'
import Pagination from '../Pagination'
import CarbonIntensityRange from '../CarbonIntensityRange/'
import { EmissionRatioResult } from '@cloud-carbon-footprint/common'

type ApexBarChartProps = {
  data: { string: [string, number] }
  dataType?: string
  emissionsData: EmissionRatioResult[]
}

const ApexBarChart: FunctionComponent<ApexBarChartProps> = ({
  data,
  dataType,
  emissionsData,
}) => {
  const [pageData, setPageData] = useState<Page<PageEntry>>({
    data: [],
    page: 0,
  })

  const theme = useTheme()
  const classes = useStyles()
  const mainTheme = theme.palette.primary.main
  const darkTheme = theme.palette.primary.dark

  let customBarColors = [mainTheme]
  if (dataType === 'region' && !!emissionsData.length) {
    customBarColors = createCustomBarColors(pageData, emissionsData, mainTheme)
  }

  const dataEntries: { x: string[]; y: number }[] = Object.entries(data)
    .filter((item: [string, [string, number]]) => item[1][1] > 0)
    .map((item) => {
      return {
        x: [item[0], `(${item[1][0]})`],
        y: item[1][1],
      }
    })
    .sort((higherC02, lowerCO2) => lowerCO2.y - higherC02.y)

  const smallestCO2E = dataEntries?.[dataEntries?.length - 1]?.y
  const largestCO2E = dataEntries?.[0]?.y
  const totalCO2EByDataType = dataEntries.reduce((acc, currentValue) => {
    return acc + currentValue.y
  }, 0)

  const pageSize = 10
  const minThreshold = 1
  const maxThreshold = 100
  const mappedDataEntries: PageEntry[] = dataEntries.map((entry) => {
    const yEntry = mapToRange(
      entry.y,
      smallestCO2E,
      largestCO2E,
      minThreshold,
      maxThreshold,
    )
    return { x: entry.x, y: isNaN(yEntry) ? maxThreshold : yEntry }
  })

  const toolbarOffset = {
    x: -120,
    y: dataType === 'region' ? -122 : -55,
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options: any = {
    series: [
      {
        name: 'Total CO2e',
        data: pageData.data,
      },
    ],
    colors: customBarColors,
    chart: {
      type: 'bar',
      toolbar: {
        offsetX: toolbarOffset.x,
        offsetY: toolbarOffset.y,
        tools: {
          download: `
            <div class="apexcharts-menu-icon" title="Menu">
                <svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true">
                <path d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"></path>
                </svg>
            </div>   
           `,
        },
      },
    },
    grid: {
      show: false,
      yaxis: {
        lines: {
          show: false,
        },
      },
      xaxis: {
        lines: {
          show: false,
        },
      },
      padding: {
        left: dataType === 'region' ? -15 : -5,
      },
    },
    plotOptions: {
      bar: {
        horizontal: true,
        barHeight: `${7 * pageData.data.length}%`,
        distributed: true,
      },
    },
    legend: {
      show: false,
    },
    dataLabels: {
      enabled: true,
      textAnchor: 'start',
      formatter: function (_: number, opts: { dataPointIndex: number }) {
        const currentCO2E =
          dataEntries[pageData.page * pageSize + opts.dataPointIndex]?.y

        const formattedPercentage = (currentCO2E / totalCO2EByDataType) * 100
        return formattedPercentage < 0.01
          ? '< 0.01 %'
          : `${formattedPercentage.toFixed(2)} %`
      },
      offsetX: 16,
      background: {
        enabled: true,
        foreColor: mainTheme,
        borderColor: darkTheme,
        padding: 6,
        borderRadius: 1,
        borderWidth: 1,
        opacity: 0.9,
      },
    },
    xaxis: {
      type: 'category',
      labels: {
        style: {
          fontSize: 0,
        },
      },
      axisBorder: {
        show: false,
      },
      max: maxThreshold,
    },
    yaxis: {
      labels: {
        minWidth: 150,
        maxWidth: 150,
        offsetY: 5,
        style: {
          fontSize: '12px',
        },
      },
    },
    tooltip: {
      fillSeriesColor: false,
      y: {
        formatter: function (value: number, opts: { dataPointIndex: number }) {
          return `${dataEntries[
            pageData.page * pageSize + opts.dataPointIndex
          ].y.toFixed(3)} metric tons`
        },
      },
    },
    height: '500px',
  }

  const handlePage = (page: Page<PageEntry>) => {
    setPageData(page)
  }

  return (
    <div className={classes.barChartContainer}>
      {dataType === 'region' && (
        <CarbonIntensityRange
          startLabel="Low carbon intensity"
          endLabel="High carbon intensity"
          colorRange={barChartCustomColors}
        />
      )}
      <Chart
        options={options}
        series={options.series}
        type="bar"
        height={options.height}
      />
      <Pagination
        data={mappedDataEntries}
        pageSize={pageSize}
        handlePage={handlePage}
      />
    </div>
  )
}

export default ApexBarChart
