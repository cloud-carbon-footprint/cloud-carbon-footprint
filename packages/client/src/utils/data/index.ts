/*
 * Â© 2021 Thoughtworks, Inc.
 */

export { fakeEmissionFactors, generateEstimations } from './generateEstimations'
export {
  mockData,
  mockDataWithUnknownsAWS,
  mockDataWithUnknownsGCP,
  mockDataWithHigherPrecision,
  mockDataWithSmallNumbers,
  mockDataWithLargeNumbers,
  mockRecommendationData,
  mockRecommendationDataWithUnknowns,
  mockEmissionsAndRecommendations,
  mockEmissionsAndRecommendationsWithUnknowns,
} from './mockData'
