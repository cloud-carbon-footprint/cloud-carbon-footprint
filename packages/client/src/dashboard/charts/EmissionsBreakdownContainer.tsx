/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { ReactElement } from 'react'
import {
  Box,
  Card,
  createStyles,
  FormControl,
  InputBase,
  makeStyles,
  MenuItem,
  Paper,
  Select,
  withStyles,
} from '@material-ui/core'
import { ChartDataTypes, EstimationResult } from '../../models/types'
import { ApexBarChart } from './ApexBarChart'

const BootstrapInput = withStyles(() =>
  createStyles({
    input: {
      border: '1px solid #ced4da',
      fontSize: 16,
      padding: '10px 26px 10px 12px',
      width: '65px',
      '&:hover': {
        borderColor: 'black',
      },
      '&:focus': {
        backgroundColor: 'white',
        borderRadius: 4,
      },
    },
  }),
)(InputBase)

const useStyles = makeStyles(() => {
  return {
    root: {
      width: '100%',
      height: '100%',
    },
    topContainer: {
      boxShadow: 'none',
      display: 'flex',
      justifyContent: 'space-between',
      alignItems: 'center',
      width: '100%',
      marginBottom: '24px',
    },
    title: {
      margin: '0',
      fontSize: '24px',
      fontFamily: 'Helvetica, Arial, sans-serif',
      opacity: '1',
      fontWeight: 900,
      color: 'rgba(0, 0, 0, 0.87)',
      padding: '.2em',
    },
  }
})

export const EmissionsBreakdownContainer = (props: { data: EstimationResult[] }): ReactElement => {
  const classes = useStyles()
  const [value, setValue] = React.useState(ChartDataTypes.REGION)

  const handleChange = (event: React.ChangeEvent<{ value: unknown }>) => {
    setValue(event.target.value as ChartDataTypes)
  }

  return (
    <Card className={classes.root}>
      <Box padding={3}>
        <Paper className={classes.topContainer}>
          <p className={classes.title}>Emissions Breakdown</p>
          <FormControl variant={'outlined'}>
            <Select value={value} onChange={handleChange} input={<BootstrapInput />}>
              <MenuItem value={ChartDataTypes.REGION}>Region</MenuItem>
              <MenuItem value={ChartDataTypes.ACCOUNT}>Account</MenuItem>
              <MenuItem value={ChartDataTypes.SERVICE}>Service</MenuItem>
            </Select>
          </FormControl>
        </Paper>
        <ApexBarChart data={props.data} dataType={value} />
      </Box>
    </Card>
  )
}
