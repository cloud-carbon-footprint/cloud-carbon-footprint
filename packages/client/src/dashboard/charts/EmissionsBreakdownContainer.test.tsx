/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import React from 'react'
import {
  create,
  ReactTestInstance,
  ReactTestRenderer,
} from 'react-test-renderer'
import { act, fireEvent, render, RenderResult } from '@testing-library/react'
import { Select } from '@material-ui/core'

import {
  EstimationResult,
  EmissionRatioResult,
} from '@cloud-carbon-footprint/common'

import { EmissionsBreakdownContainer } from './EmissionsBreakdownContainer'
import { ApexBarChart } from './ApexBarChart'
import { ServiceResult } from '../../models/types'
import useRemoteEmissionService from '../client/EmissionFactorServiceHook'
import { fakeEmissionFactors } from '../../data/generateEstimations'

jest.mock('../../themes')
jest.mock('../client/EmissionFactorServiceHook')

const mockedUseEmissionFactorService =
  useRemoteEmissionService as jest.MockedFunction<
    typeof useRemoteEmissionService
  >

describe('EmissionsBreakdownContainer', () => {
  const date1 = new Date('2020-07-10T00:00:00.000Z')
  const date2 = new Date('2020-07-11T00:00:00.000Z')
  let page: RenderResult
  let testRenderer: ReactTestRenderer, testInstance: ReactTestInstance

  const dataWithHigherPrecision: EstimationResult[] = [
    {
      timestamp: date1,
      serviceEstimates: [
        {
          cloudProvider: 'aws',
          accountName: 'testacct',
          serviceName: 'ebs',
          kilowattHours: 12.2342,
          co2e: 15.12341,
          cost: 5.82572,
          region: 'us-east-1',
          usesAverageCPUConstant: false,
        },
        {
          cloudProvider: 'aws',
          accountName: 'testacct',
          serviceName: 'ec2',
          kilowattHours: 4.745634,
          co2e: 5.234236,
          cost: 4.732,
          region: 'us-east-1',
          usesAverageCPUConstant: false,
        },
      ],
    },
    {
      timestamp: date2,
      serviceEstimates: [
        {
          cloudProvider: 'aws',
          accountName: 'testacct',
          serviceName: 'ebs',
          kilowattHours: 25.73446,
          co2e: 3.2600234,
          cost: 6.05931,
          region: 'us-east-1',
          usesAverageCPUConstant: false,
        },
        {
          cloudProvider: 'aws',
          accountName: 'testacct',
          serviceName: 'ec2',
          kilowattHours: 2.4523452,
          co2e: 7.7536,
          cost: 6.2323,
          region: 'us-east-1',
          usesAverageCPUConstant: true,
        },
      ],
    },
  ]

  beforeEach(() => {
    const mockReturnValue: ServiceResult<EmissionRatioResult> = {
      loading: false,
      data: fakeEmissionFactors,
    }
    mockedUseEmissionFactorService.mockReturnValue(mockReturnValue)
    testRenderer = create(
      <EmissionsBreakdownContainer data={dataWithHigherPrecision} />,
    )
    testInstance = testRenderer.root
    page = render(
      <EmissionsBreakdownContainer data={dataWithHigherPrecision} />,
    )
  })

  afterEach(() => {
    testRenderer.unmount()
    page.unmount()
    mockedUseEmissionFactorService.mockClear()
  })

  it('renders bar chart with dropdown', () => {
    //emulate click to test
    const allMenuItemInstancesList = testInstance.findAllByType(Select)

    expect(allMenuItemInstancesList).toHaveLength(1)

    expect(testRenderer.toJSON()).toMatchSnapshot()
  })

  it('checks to see if bar chart exists upon loading', () => {
    const isApexBarChartRendered = testInstance.findAllByType(ApexBarChart)

    expect(isApexBarChartRendered).toHaveLength(1)
  })

  it('renders emission by region bar chart by default', () => {
    const isApexBarChartRendered = testInstance.findByType(ApexBarChart)

    expect(isApexBarChartRendered.props.dataType).toBe('region')
  })

  it('selects the correct option', () => {
    act(() => {
      fireEvent.mouseDown(page.getByRole('button', { name: 'Region' }))
    })

    expect(page.getByRole('option', { name: 'Region' })).toHaveClass(
      'MuiListItem-button',
    )
    expect(page.getByRole('option', { name: 'Account' })).toHaveClass(
      'MuiListItem-button',
    )
    expect(page.getByRole('option', { name: 'Service' })).toHaveClass(
      'MuiListItem-button',
    )
  })
})
