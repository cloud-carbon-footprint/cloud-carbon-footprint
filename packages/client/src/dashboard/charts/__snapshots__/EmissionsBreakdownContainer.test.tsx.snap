// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmissionsBreakdownContainer renders donut chart with dropdown 1`] = `
<div
  className="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  id="emissionsBreakdownContainer"
>
  <div
    className="MuiBox-root MuiBox-root-4"
  >
    <div
      className="MuiPaper-root makeStyles-topContainer-2 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <p
        className="makeStyles-title-3"
      >
        Emissions Breakdown
      </p>
      <div
        className="MuiFormControl-root"
      >
        <div
          className="MuiInputBase-root MuiInputBase-formControl"
          onClick={[Function]}
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="breakdown-selector"
            className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input WithStyles(ForwardRef(InputBase))-input-5"
            id="breakdown-selector"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="button"
            tabIndex={0}
          >
            Region
          </div>
          <input
            aria-hidden={true}
            className="MuiSelect-nativeInput"
            onAnimationStart={[Function]}
            onChange={[Function]}
            required={false}
            tabIndex={-1}
            value="region"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "alignContent": "center",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
          "minHeight": "500px",
        }
      }
    >
      <div
        className="MuiPaper-root MuiCard-root makeStyles-root-6 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          className="MuiBox-root MuiBox-root-9"
          data-testid="no-data-page"
        >
          <div
            className="MuiGrid-root MuiGrid-container"
          >
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <img
                src="V1Empty-state-generic-icon.svg"
              />
              <div
                className="makeStyles-addSpacing-8"
              >
                There's no data to display!
              </div>
              <div>
                Expand your search parameters to get started.
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="makeStyles-paginationContainer-10"
      >
        <span
          style={
            Object {
              "color": "#ababab",
              "fontWeight": 700,
              "marginRight": "8px",
            }
          }
        >
          1 - 1
           of 
          1
        </span>
        <button
          aria-label="first page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="chevron-left"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="chevron-right"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="last page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
